src/components/core/GeoChart.tsx

```diff
@@ -56,89 +77,219 @@ export class GeoChartInner extends BaseVisualization<IGeoChartInnerProps, {}> {
+            execution: { executionResult, executionResponse },
         } = this.props;
-        const { size, color, segment } = geoData;
-        const sizeFormat = size ? getFormatFromExecutionResponse(size.index, executionResponse) : undefined;
 
+        if (!geoChartOptions) {
+            return {
+                config,
+                position,
+                geoData: {},
+            };
+        }
+        const { geoData } = this.geoChartOptions;
+        const { size, color, segment } = geoData;
         const { colorData, sizeData } = calculateLegendData(executionResult, geoData);
-        const colorFormat = color
-            ? getFormatFromExecutionResponse(color.index, executionResponse)
-            : undefined;
-        const isColorLegendVisible = Boolean(colorData && !segment);
+        let legendProps: IGeoChartLegendRendererProps = { config, position, geoData };
 
-        const legendProps = {
-            config,
-            position,
-            sizeData,
-            sizeFormat,
-            colorData,
-            colorFormat,
-            isColorLegendVisible,
-        };
-        return legendRenderer(legendProps);
+        if (sizeData) {
+            const sizeLegendOptions = {
+                sizeData,
+                sizeFormat: getFormatFromExecutionResponse(size.index, executionResponse),
+            };
+            legendProps = { ...legendProps, ...sizeLegendOptions };
+        }
+
+        if (colorData) {
+            const colorLegendOptions = {
+                colorData,
+                colorFormat: getFormatFromExecutionResponse(color.index, executionResponse),
+            };
+            legendProps = { ...legendProps, ...colorLegendOptions };
+        }
+
+        if (segment && this.state.enabledLegendItems.length) {
+            const segmentLegendOptions = {
+                segmentData: this.state.enabledLegendItems,
+                onItemClick: this.onLegendItemClick,
+            };
+            legendProps = { ...legendProps, ...segmentLegendOptions };
+        }
+
+        return legendProps;
+    }
+
+    private getChartProps(): IGeoChartRendererProps {
+        const { geoChartOptions } = this;
+        const { config, execution, afterRender } = this.props;
+
+        if (!geoChartOptions) {
+            return {
+                config,
+                execution,
+                afterRender,
+                geoData: {},
+            };
+        }
+
+        const { geoData } = geoChartOptions;
+        const segmentIndex: number = get(geoChartOptions, "geoData.segment.index");
+        let chartProps: IGeoChartRendererProps = { config, execution, afterRender, geoData };
+
+        if (segmentIndex) {
+            const selectedSegmentNames: string[] = this.state.enabledLegendItems.reduce(
+                (result: string[], item: IPushpinCategoryLegendItem): string[] => {
+                    if (item.isVisible) {
+                        result.push(item.name);
```

Comment:
I prefer to use `spread operator` than `push` in this case.
https://gooddata.slack.com/archives/GSQP9FDU5/p1579232966009400

Reply:
I will change, but I don't see any benefit for this case.